# Genome and population sequencing of aspen advances Populus as a model system for plant research 

## Authors
**Yao-Cheng Lin#1,2, Jing Wang#3,4, Nicolas Delhomme#5, Bastian Shiffthaler6, Görel Sundström5,7, Andrea Zuccolo8,9, Björn Nystedt10, Torgeir R. Hvidsten6,11, Amanda de la Torre3,12, Rosa Cossu9, Marc Hoppner13,14 , Henrik Lantz13,15, Douglas G. Scofield3,16,17, Neda Zamani6,13, Anna Johansson10, Chanaka Mannapperuma6, Kathryn M. Robinson6, Niklas Mähler6, Ilia Leitch18, Jaume Pellicer18, Eung-Jun Park19, Yves Van der Peer1,20, Manfred Grabherr13,14, Stefan Jansson6, Pär K. Ingvarsson3,21, Nathaniel R. Street*6**

\# equal contribution

## Affiliations

1. Department of Plant Systems Biology (VIB) and Department of Plant Biotechnology and Bioinformatics (Ghent University), Ghent, Belgium
2. Biotechnology Center in Southern Taiwan and Agricultural Biotechnology Research Center, Academia Sinica, Tainan, Taiwan
3. Umeå Plant Science Centre, Department of Ecology and Environmental Science, Umeå University, Umeå, Sweden
4. Centre for Integrative Genetics, Faculty of Biosciences, Norwegian University of Life Sciences, Ås, Norway 
5. Umeå Plant Science Centre, Department of Forest Genetics and Plant Physiology, Swedish University of Agricultural Sciences, Umeå, Sweden
6. Umeå Plant Science Centre, Department of Plant Physiology, Umeå University, Umeå, Sweden
7. Umeå Plant Science Centre, Department of Forest Genetics and Plant Physiology, SLU, SE-901 85, Umeå, Sweden
8. Istituto di Genomica Applicata, Via J. Linussio 51, 33100 Udine, Italy
9. Institute of Life Sciences, Scuola Superiore Sant’Anna, 56127 Pisa, Italy
10. Wallenberg Advanced Bioinformatics Infrastructure, Science for Life Laboratory, Uppsala University, Uppsala, Sweden
11. Department of Chemistry, Biotechnology and Food Science, Norwegian University of Life Sciences, Ås, Norway
12. Department of Plant Sciences, University of California, Davis, California, USA
13. Department of Medical Biochemistry and Microbiology, Uppsala University, Uppsala 
14. Institut für Klinische Molekularbiologie, Christian-Albrechts-Universität zu Kiel, Kiel, Germany
15. National Bioinformatics Infrastructure Sweden, Uppsala University, Uppsala
16. Department of Ecology and Genetics: Evolutionary Biology, Uppsala University, Uppsala, Sweden
17. Uppsala Multidisciplinary Center for Advanced Computational Science, Uppsala University, Uppsala, Sweden
18. Character Evolution team, Department of Comparative Plant and Fungal Biology, Royal Botanic Gardens, Kew, England
19. Division of Forest Biotechnology, Korea Forest Research Institute, Suwon, Korea
20. Genomics Research Institute, University of Pretoria, Pretoria, South Africa
21. Department of Plant Biology, Swedish University of Agricultural Sciences, Uppsala, Sweden

## Abstract

Aspens include European trembling (_Populus tremula_) and north American quaking (_P. tremuloides_) aspen, European white aspen (_P. alba_) and numerous Asian species. Aspens spread as clonal colonies that can persist for millennia [1], are keystone species supporting a wide diversity of species [2] and produce a diverse array of secondary metabolites. Here, we present genome and population genetics characterisation of _P. tremula_ and _P. tremuloides_. Our analyses show that while the aspen genome structure is similar to that of _P. trichocarpa_ reference genome [3], with highly conserved gene content, they are considerably more heterozygous. One notable exception is the sex-determination region, which has altered gene content compared to _P. trichocarpa_. We used population re-sequencing data to examine the relationship between gene characteristics, genomic context and evidence of selection, finding that 5’ UTRs have high rates of positive and negative selection and confirming a negative relationship between gene co-expression connectivity and purifying selection. The aspen genomics resources presented here greatly improve the utility of Populus as a model system for functional and evolutionary genomics. Data collections and analysis tools related to this study are available at [PopGenIE](http://popgenie.org) [4].

## Content of this repository

This repository contains the custom scripts mentioned in the supplementary file of the manuscript. These are sorted in directories corresponding to the title of the sections they are mentioned in.

## References

1.	Stettler, R. F. & Bradshaw Jr, H. D. in Biology of Populus and its implications for management and conservation (eds. Stettler, R. ., Bradshaw Jr, H. D., Heilman, P. & Hinckley, T.) 1–6 (NRC Research Press, 1996).
2.	Robinson, K., Ingvarsson, P., Jansson, S. & Albrectsen, B. Genetic Variation in Functional Traits Influences Arthropod Community Composition in Aspen (Populus tremula L.). PLoS One 7, e37679 (2012).
3.	Tuskan, G. A. et al. The genome of black cottonwood, Populus trichocarpa (Torr. & Gray). Science 313, 1596–604 (2006).
4.	Sundell, D. et al. The Plant Genome Integrative Explorer Resource: PlantGenIE.org. New Phytol. 208, 1149–1156 (2015).
